File: kotlin/src/test/kotlin/com/bhengubv/circleai/family/FamilyTest.kt
// FamilyTest.kt
//
// Verifies the CircleAI.Family port against the C# reference:
//   - add/get; Members ordered by name; eventsForMember filters by MemberIds
//     membership, ordered by AtUtc; record + totalPaidBy / spendByCategory with
//     since filter (category case-insensitive)
//   - domain-context constants; adapter enrichment + a domain helper

package com.bhengubv.circleai.family

import com.bhengubv.circleai.companion.support.FakeCompanionSession
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import java.math.BigDecimal
import java.time.Instant
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class FamilyTest {

    private fun assertMoney(expected: String, actual: BigDecimal) =
        assertTrue(BigDecimal(expected).compareTo(actual) == 0, "expected $expected but was $actual")

    private val t0 = Instant.parse("2026-07-01T00:00:00Z")

    @Test
    fun `members events and expenses`() {
        val b = InMemoryFamilyBoard()
        b.add(FamilyMember("m2", "Zoe", "child", t0))
        b.add(FamilyMember("m1", "Ann", "parent", t0))
        assertEquals(listOf("Ann", "Zoe"), b.members.map { it.name }) // ordered
        assertEquals("Zoe", b.getMember("m2")!!.name)

        b.schedule(FamilyEvent("e1", "Dinner", t0.plusSeconds(7200), listOf("m1", "m2")))
        b.schedule(FamilyEvent("e2", "Soccer", t0.plusSeconds(3600), listOf("m2"))) // earlier
        b.schedule(FamilyEvent("e3", "Work", t0, listOf("m1")))
        // events for m2 ordered by AtUtc ASC: e2 (earlier) before e1
        assertEquals(listOf("e2", "e1"), b.eventsForMember("m2").map { it.eventId })

        b.record(SharedExpense("x1", "m1", BigDecimal("300"), "ZAR", "Food", t0.plusSeconds(10)))
        b.record(SharedExpense("x2", "m1", BigDecimal("150"), "ZAR", "food", t0.plusSeconds(20)))
        b.record(SharedExpense("x3", "m2", BigDecimal("50"), "ZAR", "Toys", t0.minusSeconds(9999))) // before since
        val since = t0
        assertMoney("450", b.totalPaidBy("m1", since))      // 300 + 150
        assertMoney("450", b.spendByCategory("FOOD", since)) // case-insensitive, both food
        assertMoney("0", b.totalPaidBy("m2", since))         // m2's only expense predates since
    }

    @Test
    fun `domain context and adapter`() = runTest {
        assertTrue(FamilyDomainContext.SYSTEM_PROMPT_SNIPPET.startsWith("[DOMAIN: Family]"))
        assertTrue("Childrens_Act_38_2005" in FamilyDomainContext.complianceFlags)

        val fake = FakeCompanionSession()
        val a = FamilyCompanionAdapter(fake)
        a.sendAsync("hi")
        assertTrue(fake.lastMessage!!.startsWith("[DOMAIN: Family]"))
        a.mediateSiblingDisputeAsync("6 and 9", "toy sharing")
        assertTrue(fake.lastMessage!!.contains("sibling dispute"))
    }
}

File: rust/tests/family_test.rs
//! family_test.rs
//!
//! Ports the behaviour of `CircleAI.Family`: member registry (name-ordered),
//! shared events per member (time-ordered), and shared-expense totals by member
//! and by category (since a cutoff).

use chrono::{Duration, Utc};
use circle_ai::family::{
    FamilyEvent, FamilyMember, IFamilyBoard, InMemoryFamilyBoard, SharedExpense,
};

#[test]
fn members_added_and_name_ordered() {
    let board = InMemoryFamilyBoard::new();
    assert!(board.get_member("m1").is_none());
    board.add(FamilyMember::new("m2", "Zed", "Child", Utc::now()));
    board.add(FamilyMember::new("m1", "Amy", "Parent", Utc::now()));
    let members = board.members();
    let names: Vec<&str> = members.iter().map(|m| m.name.as_str()).collect();
    assert_eq!(names, vec!["Amy", "Zed"]);
}

#[test]
fn events_for_member_time_ordered() {
    let board = InMemoryFamilyBoard::new();
    let now = Utc::now();
    board.schedule(FamilyEvent::new("e1", "Dinner", now + Duration::hours(3), vec!["m1".into(), "m2".into()]));
    board.schedule(FamilyEvent::new("e2", "Doctor", now + Duration::hours(1), vec!["m1".into()]));
    board.schedule(FamilyEvent::new("e3", "Soccer", now + Duration::hours(2), vec!["m2".into()]));

    let for_m1 = board.events_for_member("m1");
    let ids: Vec<&str> = for_m1.iter().map(|e| e.event_id.as_str()).collect();
    assert_eq!(ids, vec!["e2", "e1"]); // time-ordered, only events including m1
}

#[test]
fn expense_totals_by_member_and_category() {
    let board = InMemoryFamilyBoard::new();
    let now = Utc::now();
    let since = now - Duration::days(7);
    board.record(SharedExpense::new("x1", "m1", 100.0, "USD", "Groceries", now));
    board.record(SharedExpense::new("x2", "m1", 50.0, "USD", "groceries", now - Duration::days(1)));
    board.record(SharedExpense::new("x3", "m2", 30.0, "USD", "Fuel", now));
    board.record(SharedExpense::new("x4", "m1", 999.0, "USD", "Groceries", now - Duration::days(30))); // before cutoff

    assert_eq!(board.total_paid_by("m1", since), 150.0);
    assert_eq!(board.spend_by_category("GROCERIES", since), 150.0);
    assert_eq!(board.spend_by_category("Fuel", since), 30.0);
}
